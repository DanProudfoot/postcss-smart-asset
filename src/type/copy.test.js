import { processedCss, compareFixtures } from "../../test/setup"

describe("copy without assetsPath", () => {
  const opts = {
    url: "copy"
  }

  compareFixtures("cant-copy", "shouldn't copy assets if not info available", opts)

  const postcssOpts = {
    from: "test/fixtures/index.css",
    to: "test/fixtures/build/index.css"
  }

  testCopy(opts, postcssOpts)
})

describe("copy with assetsPath", () => {
  const opts = {
    url: "copy",
    assetsPath: "assets"
  }

  compareFixtures("cant-copy", "shouldn't copy assets if not info available", opts)

  const postcssOpts = {
    from: "test/fixtures/index.css",
    to: "test/fixtures/build/index.css"
  }

  testCopy(opts, postcssOpts)
})

describe('copy with assetsPath and without "to"', () => {
  const opts = {
    url: "copy",
    assetsPath: "test/fixtures/build/assets"
  }

  const postcssOpts = {
    from: "test/fixtures/index.css"
  }

  testCopy(opts, postcssOpts)
})

describe("copy when inline fallback", () => {
  const opts = {
    url: "inline",
    maxSize: 0.0001,
    fallback: "copy",
    assetsPath: "assets"
  }

  compareFixtures("cant-copy", "shouldn't copy assets if not info available", opts)

  const postcssOpts = {
    from: "test/fixtures/index.css",
    to: "test/fixtures/build/index.css"
  }

  testCopy(opts, postcssOpts)
})

function testCopy(opts, postcssOpts) {
  const optsWithHash = Object.assign({}, opts, { useHash: true })
  const assetsPath = opts.assetsPath ? `${opts.assetsPath}\/` : ""
  const patterns = {
    copyPixelPng: new RegExp(`"${assetsPath}imported\/pixel\.png"`),
    copyPixelGif: new RegExp(`"${assetsPath}pixel\\.gif"`),
    copyParamsPixelPngHash: new RegExp(`"${assetsPath}imported\/pixel\\.png\\?\#iefix"`),
    copyParamsPixelPngParam: new RegExp(`"${assetsPath}imported\/pixel\\.png\\?foo=bar"`),
    copyParamsPixelGif: new RegExp(`"${assetsPath}pixel\\.gif\\#el"`),
    copyXXHashPixel8: new RegExp(`"${assetsPath}[a-z0-9]{8}\\.png"`),
    copyXXHashParamsPixel8: new RegExp(`"${assetsPath}[a-z0-9]{8}\\.png\\?v=1\\.1\\#iefix"`)
  }
  const matchAll = (css, patternsKeys) =>
    expect(patternsKeys.every((pat) => css.match(patterns[pat]))).toBeTruthy()

  describe("should copy asset from the source (`from`) to the assets destination (`to` + `assetsPath`)", () => {
    test("rebase the url", () => {
      return processedCss("copy", opts, postcssOpts).then((css) => {
        matchAll(css, [ "copyPixelPng", "copyPixelGif" ])
      })
    })

    test("rebase the url keeping parameters", () => {
      return processedCss("copy-parameters", opts, postcssOpts).then((css) => {
        matchAll(css, [ "copyParamsPixelPngHash", "copyParamsPixelPngParam", "copyParamsPixelGif" ])
      })
    })

    test("rebase the url using a hash name", () => {
      return processedCss("copy-hash", optsWithHash, postcssOpts).then((css) => {
        matchAll(css, [ "copyXXHashPixel8" ])
      })
    })

    test("rebase the url using a hash name keeping parameters", () => {
      return processedCss("copy-hash-parameters", optsWithHash, postcssOpts).then((css) => {
        matchAll(css, [ "copyXXHashParamsPixel8" ])
      })
    })
  })
}
