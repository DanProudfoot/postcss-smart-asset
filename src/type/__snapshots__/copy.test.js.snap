// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`copy when inline fallback should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url 1`] = `
"body {
  clip-path: url(\\"./assets/imported/pixel.png\\")
}

div {
  clip-path: url(\\"./assets/pixel.gif\\")
}"
`;

exports[`copy when inline fallback should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url appending a hash to the name 1`] = `
"body {
  clip-path: url(\\"./assets/pixel_fbDRJoaV.png\\")
}"
`;

exports[`copy when inline fallback should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url keeping parameters 1`] = `
"body {
  clip-path: url(\\"./assets/imported/pixel.png?#iefix\\")
}

header {
  clip-path: url(\\"./assets/imported/pixel.png?foo=bar\\")
}

div {
  clip-path: url(\\"./assets/pixel.gif#el\\")
}

span {
  clip-path: url(\\"./assets/pixel.svg#el\\")
}"
`;

exports[`copy when inline fallback should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url using a hash name 1`] = `
"body {
  clip-path: url(\\"./assets/fbDRJoaV.png\\")
}"
`;

exports[`copy when inline fallback should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url using a hash name keeping parameters 1`] = `
"body {
  clip-path: url(\\"./assets/fbDRJoaV.png?v=1.1#iefix\\")
}"
`;

exports[`copy when inline fallback should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url using appending a hash to the name name keeping parameters 1`] = `
"body {
  clip-path: url(\\"./assets/pixel_fbDRJoaV.png?v=1.1#iefix\\")
}"
`;

exports[`copy when inline fallback shouldn't copy assets if not info available 1`] = `
"body {
  background: url(\\"./one\\");
}"
`;

exports[`copy with assetsPath and without "to" should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url 1`] = `
"body {
  clip-path: url(\\"./test/fixtures/build/assets/imported/pixel.png\\")
}

div {
  clip-path: url(\\"./test/fixtures/build/assets/pixel.gif\\")
}"
`;

exports[`copy with assetsPath and without "to" should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url appending a hash to the name 1`] = `
"body {
  clip-path: url(\\"./test/fixtures/build/assets/pixel_fbDRJoaV.png\\")
}"
`;

exports[`copy with assetsPath and without "to" should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url keeping parameters 1`] = `
"body {
  clip-path: url(\\"./test/fixtures/build/assets/imported/pixel.png?#iefix\\")
}

header {
  clip-path: url(\\"./test/fixtures/build/assets/imported/pixel.png?foo=bar\\")
}

div {
  clip-path: url(\\"./test/fixtures/build/assets/pixel.gif#el\\")
}

span {
  clip-path: url(\\"./test/fixtures/build/assets/pixel.svg#el\\")
}"
`;

exports[`copy with assetsPath and without "to" should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url using a hash name 1`] = `
"body {
  clip-path: url(\\"./test/fixtures/build/assets/fbDRJoaV.png\\")
}"
`;

exports[`copy with assetsPath and without "to" should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url using a hash name keeping parameters 1`] = `
"body {
  clip-path: url(\\"./test/fixtures/build/assets/fbDRJoaV.png?v=1.1#iefix\\")
}"
`;

exports[`copy with assetsPath and without "to" should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url using appending a hash to the name name keeping parameters 1`] = `
"body {
  clip-path: url(\\"./test/fixtures/build/assets/pixel_fbDRJoaV.png?v=1.1#iefix\\")
}"
`;

exports[`copy with assetsPath should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url 1`] = `
"body {
  clip-path: url(\\"./assets/imported/pixel.png\\")
}

div {
  clip-path: url(\\"./assets/pixel.gif\\")
}"
`;

exports[`copy with assetsPath should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url appending a hash to the name 1`] = `
"body {
  clip-path: url(\\"./assets/pixel_fbDRJoaV.png\\")
}"
`;

exports[`copy with assetsPath should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url keeping parameters 1`] = `
"body {
  clip-path: url(\\"./assets/imported/pixel.png?#iefix\\")
}

header {
  clip-path: url(\\"./assets/imported/pixel.png?foo=bar\\")
}

div {
  clip-path: url(\\"./assets/pixel.gif#el\\")
}

span {
  clip-path: url(\\"./assets/pixel.svg#el\\")
}"
`;

exports[`copy with assetsPath should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url using a hash name 1`] = `
"body {
  clip-path: url(\\"./assets/fbDRJoaV.png\\")
}"
`;

exports[`copy with assetsPath should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url using a hash name keeping parameters 1`] = `
"body {
  clip-path: url(\\"./assets/fbDRJoaV.png?v=1.1#iefix\\")
}"
`;

exports[`copy with assetsPath should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url using appending a hash to the name name keeping parameters 1`] = `
"body {
  clip-path: url(\\"./assets/pixel_fbDRJoaV.png?v=1.1#iefix\\")
}"
`;

exports[`copy with assetsPath shouldn't copy assets if not info available 1`] = `
"body {
  background: url(\\"./one\\");
}"
`;

exports[`copy without assetsPath should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url 1`] = `
"body {
  clip-path: url(\\"./imported/pixel.png\\")
}

div {
  clip-path: url(\\"./pixel.gif\\")
}"
`;

exports[`copy without assetsPath should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url appending a hash to the name 1`] = `
"body {
  clip-path: url(\\"./pixel_fbDRJoaV.png\\")
}"
`;

exports[`copy without assetsPath should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url keeping parameters 1`] = `
"body {
  clip-path: url(\\"./imported/pixel.png?#iefix\\")
}

header {
  clip-path: url(\\"./imported/pixel.png?foo=bar\\")
}

div {
  clip-path: url(\\"./pixel.gif#el\\")
}

span {
  clip-path: url(\\"./pixel.svg#el\\")
}"
`;

exports[`copy without assetsPath should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url using a hash name 1`] = `
"body {
  clip-path: url(\\"./fbDRJoaV.png\\")
}"
`;

exports[`copy without assetsPath should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url using a hash name keeping parameters 1`] = `
"body {
  clip-path: url(\\"./fbDRJoaV.png?v=1.1#iefix\\")
}"
`;

exports[`copy without assetsPath should copy asset from the source (\`from\`) to the assets destination (\`to\` + \`assetsPath\`) rebase the url using appending a hash to the name name keeping parameters 1`] = `
"body {
  clip-path: url(\\"./pixel_fbDRJoaV.png?v=1.1#iefix\\")
}"
`;

exports[`copy without assetsPath shouldn't copy assets if not info available 1`] = `
"body {
  background: url(\\"./one\\");
}"
`;
